File: Leetcode/C++/151. Reverse Words in a String/solution.cpp
// https://leetcode.com/problems/reverse-words-in-a-string/
// Ref: https://leetcode.com/problems/reverse-words-in-a-string/description/?envType=study-plan-v2&envId=top-interview-150

#include <string>

using namespace std;

class Solution
{
public:
    string reverseWords(string s)
    {
        reverse(s.begin(), s.end());
        // left, right = start, end index of word
        int left = 0, right = 0, index = 0;
        while (index < s.size())
        {
            // Find not empty word
            while (index < s.size() && s[index] != ' ')
            {
                s[right++] = s[index++];
            }
            // If find not empty word reverse it
            if (left < right)
            {
                reverse(s.begin() + left, s.begin() + right);
                if (right == s.size())
                    break;
                s[right++] = ' '; // Insert blank between word
                left = right;
            }

            index++;
        }
        // Delete last empty char
        if (right > 0 && s[right - 1] == ' ')
            --right;
        s.resize(right);
        return s;
    }
};

File: Leetcode/Python/151. Reverse Words in a String/solution.py
# https://leetcode.com/problems/reverse-words-in-a-string/

# Python string is immutable so need O(N) memory

class Solution:
    def reverseWords(self, s: str) -> str:
        words = s.split()

        left, right = 0, len(words) - 1
        while left < right:
            words[left], words[right] = words[right], words[left]
            left += 1
            right -= 1

        return " ".join(words)


class Solution:
    def reverseWords(self, s: str) -> str:
        return " ".join(reversed(s.split()))


class Solution:
    def reverseWords(self, s: str) -> str:
        idx = 0

        # Split word
        words = []
        temp = []
        while idx < len(s):
            if s[idx] == " ":
                if not temp:
                    idx += 1
                    continue
                words.append("".join(temp))
                temp.clear()
            else:
                temp.append(s[idx])
            idx += 1

        if temp:
            words.append("".join(temp))

        words.reverse()
        return " ".join(words)
